### PR TITLE
Add decoupled weight decay factory to create optimizerWs.

### PR DESCRIPTION
**Pull request by PhilJd**
As discussed in #22343, this pull request adds a factory function to create weight decay optimizers ([Loshchilov & Hutter](https://arxiv.org/abs/1711.05101)), reducing code duplication and allowing to extend arbitrary optimizers with decoupled weight decay.

Includes AdamW and SGDW by default.

Note:  This implementation does not multiply by learning rate and is therefore blocked by #22343.

Note: The current implementation only works for optimizers that use the gradient but not the value of the parameter to optimize as the weight decay is applied before the optimizer update. If desired, I can extend the factory with a  boolean `optimizer_requires_value`, which forces decay of the variable as the last step (requiring a copy of the parameter).

Ping @vincentqb :)

### CODE REVIEW

File: docs/source/optim.rst [hunk 1]
```diff
@@ -113,6 +113,11 @@ Algorithms
     :members:
 .. autoclass:: AdamW
     :members:
+.. autoclass:: AdamW2
```
Comment (vincentqb) [inline docs/source/optim.rst]: What is AdamW2?

There is an implementation of AdamW in PyTorch since #21250. In the version implemented, the decay is taken from theta, but the weight and lr schedulers are still coupled. I would expect the new implementation to supersede the current one.

File: torch/optim/decoupled_weight_decay_optim.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            **kwargs: keyword arguments passed to the base optimizer.
+        
+        Details on the update step can be found in the documentation of the
+        wrapped {0} optimizer.
+        """.format(base_optimizer.__name__)
+
+        def __init__(self, decoupled_weight_decay, *args, **kwargs):
+            super(DecoupledWeightDecayOptimizer, self).__init__(*args, **kwargs)
+            if self.defaults["weight_decay"] != 0:
+                warnings.warn(
```
Comment (vincentqb) [inline torch/optim/decoupled_weight_decay_optim.py]: Do people want to do both at the same time? I suspect people understand AdamW and SGDW as not having L2 anymore. If so, I'd suggest to simply biggy back on the parameter provided in Adam. There shouldn't be confusion when someone invokes AdamW which regularization they want.

File: torch/optim/decoupled_weight_decay_optim.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+        def step(self, closure=None):
+            # First perform the weight decay, then execute the optimizer step.
+            # This way, we don't need to store the parameter value prior to the
+            # update. The drawback is, that the optimizer's step function must
+            # not depend on the parameter value but only the gradient.
+            for group in self.param_groups:
+                for p in group['params']:
+                    if p.grad is not None:
+                        p.data.mul_(1 - group['decoupled_weight_decay'])
```
Comment (vincentqb) [inline torch/optim/decoupled_weight_decay_optim.py]: We'll need tests here showing under which condition this implementation and the current one agree.

**Review by vincentqb — COMMENTED** [review]
Thanks for looking into this!

> As discussed in #22343, this pull request adds a factory function to create weight decay optimizers ([Loshchilov & Hutter](https://arxiv.org/abs/1711.05101)), reducing code duplication and allowing to extend arbitrary optimizers with decoupled weight decay.

The duplication of code between SGD/W and Adam/W is not necessarily bad -- and the versions are very easy to compare with a simple diff. However, I do agree that, if more and more optimizers want the weight decay, it would not be sustainable. We also do want a standard way of introducing weight (or other parameters) schedulers (and I'll link back [here](https://github.com/pytorch/pytorch/issues/22343#issuecomment-513003798) to reference back on schedulers' design).

> Includes AdamW and SGDW by default.
> 
> Note: This implementation does not multiply by learning rate and is therefore blocked by #22343.

We will need to make sure the current AdamW from #21250 and this new version agree under reasonable conditions, so we can supersede the current version with this one.
 
> Note: The current implementation only works for optimizers that use the gradient but not the value of the parameter to optimize as the weight decay is applied before the optimizer update. If desired, I can extend the factory with a boolean `optimizer_requires_value`, which forces decay of the variable as the last step (requiring a copy of the parameter).

I would indeed expect a weight scheduler to be applied at the end. Which step makes you require the copy of the parameter when moving the weight decay at the end?

File: docs/source/optim.rst (hunk 1 above)
Comment (PhilJd) [reply docs/source/optim.rst]: Sorry, AdamW2 was a relict from comparing the old with the new optimizer. 
I've now copied the same update step to the wrapper.

File: torch/optim/decoupled_weight_decay_optim.py (hunk 2 above)
Comment (PhilJd) [reply torch/optim/decoupled_weight_decay_optim.py]: I've changed the the input argument to weight_decay. Internally however, it still needs to be called differently, otherwise the optimizers' update steps will do additional L2 decay.

File: torch/optim/decoupled_weight_decay_optim.py (hunk 3 above)
Comment (PhilJd) [reply torch/optim/decoupled_weight_decay_optim.py]: This I haven't included yet, what's the format you suggest? Keep the old optimizer and compare the outputs?

**Comment by PhilJd** [comment]
Thanks for the review :)
I'm decaying first, as the update step usually only requires the gradient (so changing p has no effect), while the decay depends on the value of the parameter (so changing p beforehand has an effect).

- original formula:
 `p = -(lr *  p.grad) - (p * wd)     # p = 1.8`
- decay after update, uses the update p for decay:
  `opt.step()    # p -> 1`
  `decay(p)    # p -> 0.9`
- decay before update uses the original value of p.
  `decay(p)   # p -> 1.8 `
  `opt.step()  # p -> 1`


Optimizers that rely on the value of the parameter (or on closures) will still need a custom implementation of decoupled weight decay. This allows to do the copy within the loop over the parameters, only copying one parameter at a time, compared to copying all parameters in a wrapper.
However, most commonly used optimizers only use the gradient, so this wrapper is still reasonable ;)

File: torch/optim/decoupled_weight_decay_optim.py (hunk 2 above)
Comment (PhilJd) [reply torch/optim/decoupled_weight_decay_optim.py]: On a second thought, if general schedules are on their way, it might be confusing having to schedule "decoupled_weight_decay", while the input argument is called "weight decay"?
